### PR TITLE
Closes #1729 derive_expected_records

### LINKED ISSUE #1729: Feature Request:  Implement `derive_expected_records()`
### Feature Idea

```r
derive_expected_records(
    dataset,
    dataset_expected_obs, 
    by_vars,          
    set_values_to      
    )
```
The function adds expected records to a dataset based on `dataset_expected_obs` (same argument as in `derive_locf_records`).
I labeled `ADQS` as it was initially discussed in `ADQS` workflow https://github.com/pharmaverse/admiral/issues/1280, but this function applies to other datasets as well.

### Relevant Input

_No response_

### Relevant Output

_No response_

### Reproducible Example/Pseudo Code

_No response_

### PR DESCRIPTION
**Pull request by kaz462**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiraldev/devel/articles/development_process.html) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `devel` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/devel/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/devel/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/devel/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately 
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

**Comment by kaz462** [comment]
@bundfussr, I got below error message under `devel` as mentioned in https://github.com/pharmaverse/admiral/pull/1645
```r
> devtools::load_all()
ℹ Loading admiral
Error in assert_filter_cond(enexpr(filter), optional = TRUE) : 
  is_quosure(arg) is not TRUE
In addition: Warning message:
replacing previous import ‘admiraldev::as_name’ by ‘rlang::as_name’ when loading ‘admiral’ 
```

```
$ git status
On branch devel
Your branch is up to date with 'origin/devel'.

nothing to commit, working tree clean
```

**Comment by bundfussr** [comment]
@kaz462 , are you using the latest devel version of admiraldev?

**Comment by kaz462** [comment]
ah, thanks @bundfussr ! it's working now

File: tests/testthat/test-derive_expected_records.R [hunk 1]
```diff
@@ -0,0 +1,43 @@
+## Test 1: `by_vars` are missing ----
+test_that("derive_expected_records Test 1: `by_vars` are missing", {
```
Comment (bundfussr) [inline tests/testthat/test-derive_expected_records.R:2]: The title does not match the code.
Could we have two tests, one with `by_vars` missing and one with `by_vars` not missing?

**Review by bundfussr — CHANGES_REQUESTED** [review]

File: tests/testthat/test-derive_expected_records.R (hunk 1 above)
Comment (kaz462) [reply tests/testthat/test-derive_expected_records.R:2]: how about the below four sets? 
1. missing values in `by_vars`
2. `by_vars` = `NULL`
3. visit variables are parameter independent
4. visit variables are parameter dependent

File: R/derive_expected_records.R [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+#' library(tibble)
+#'
+#' adqs <- tribble(
+#'   ~USUBJID, ~PARAMCD, ~AVISITN, ~AVISIT, ~AVAL,
+#'   "1", "a", 1, "WEEK 1", 10,
+#'   "1", "b", 2, "WEEK 2", 11,
+#'   "2", "a", 1, "WEEK 1", 12,
+#'   "2", "a", 2, "WEEK 2", 13,
+#'   "2", "b", 2, "WEEK 2", 14,
+#' )
```
[the remark this answers is not in the document]
Comment (kaz462) [reply R/derive_expected_records.R:57]: should I manually align the columns? how can I pass the code style workflow after aligning the columns?

[the remark this answers is not in the document]
Comment (bundfussr) [reply R/derive_expected_records.R]: Unfortunately, you have to align the columns manually. Maybe we should have a plugin for this. I'll set it on the agenda.
If _all_ columns are left or right aligned, styler should preserve the alignment. Usually I left align character columns and right align numeric columns.

File: tests/testthat/test-derive_expected_records.R (hunk 1 above)
Comment (bundfussr) [reply tests/testthat/test-derive_expected_records.R]: Looks good.

Comment (kaz462) [reply tests/testthat/test-derive_expected_records.R]: the second case `by_vars = NULL` failed in one of the 3 ubuntu checks. I couldn't figure out why, any suggestions on how to reproduce the error? I tried to use the same R version `4.0.5` but the unit tests are all passed in my end.

Comment (bundfussr) [reply tests/testthat/test-derive_expected_records.R]: Hm, that's tricky.
You could try the code stated in https://pharmaverse.github.io/admiraldev/devel/articles/pr_review_guidance.html#check-fails-only-on-one-version to reproduce the error.

Comment (kaz462) [reply tests/testthat/test-derive_expected_records.R]: Thanks, @bundfussr for the document. I was able to reproduce the error and locate the issue (`tidyr::crossing`), but I'm not sure how to fix this. 
@cicdguy @bms63 would you help check?
below is the issue example from testing environment based on [Check Fails Only on One Version](https://pharmaverse.github.io/admiraldev/devel/articles/pr_review_guidance.html#check-fails-only-on-one-version)
```r
> library(dplyr)
> ids <- tibble::tibble(x=1)[, 0] 
> data_expected <- tibble::tribble(
+   ~PARAMCD, ~AVISITN, ~AVISIT,
+   "a", 1, "WEEK 1",
+   "a", 2, "WEEK 2",
+   "b", 1, "WEEK 1",
+   "b", 2, "WEEK 2"
+ )
> ids %>%
+   tidyr::crossing(data_expected)
# A tibble: 0 x 3
# … with 3 variables: PARAMCD <chr>, AVISITN <dbl>, AVISIT <chr>
```

```r
> sessionInfo()
R version 4.0.5 (2021-03-31)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/atlas/libblas.so.3.10.3
LAPACK: /usr/lib/x86_64-linux-gnu/atlas/liblapack.so.3.10.3

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8        LC_COLLATE=C.UTF-8    
 [5] LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8    LC_PAPER=C.UTF-8       LC_NAME=C             
 [9] LC_ADDRESS=C           LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] dplyr_1.0.5

loaded via a namespace (and not attached):
 [1] pillar_1.5.1      compiler_4.0.5    highr_0.8         remotes_2.2.0     tools_4.0.5       digest_0.6.27    
 [7] evaluate_0.14     lifecycle_1.0.0   tibble_3.1.0      pkgconfig_2.0.3   rlang_0.4.10      reprex_1.0.0     
[13] cli_2.3.1         rstudioapi_0.14   curl_4.3          yaml_2.2.1        xfun_0.22         withr_2.4.1      
[19] styler_1.4.1      knitr_1.31        generics_0.1.0    fs_1.5.0          vctrs_0.3.7       tidyselect_1.1.0 
[25] glue_1.4.2        R6_2.5.0          processx_3.5.0    fansi_0.4.2       rmarkdown_2.7     tidyr_1.1.3      
[31] purrr_0.3.4       callr_3.6.0       clipr_0.7.1       magrittr_2.0.1    ps_1.6.0          backports_1.2.1  
[37] ellipsis_0.3.1    htmltools_0.5.1.1 assertthat_0.2.1  utf8_1.2.1        crayon_1.4.1 
```

Same environment but using `renv` produces the right results -
```r
> library(dplyr)
> ids <- tibble::tibble(x=1)[, 0] 
> data_expected <- tibble::tribble(
  +   ~PARAMCD, ~AVISITN, ~AVISIT,
  +   "a", 1, "WEEK 1",
  +   "a", 2, "WEEK 2",
  +   "b", 1, "WEEK 1",
  +   "b", 2, "WEEK 2"
  + )
> ids %>%
  +   tidyr::crossing(data_expected)
# A tibble: 4 x 3
PARAMCD AVISITN AVISIT
<chr>     <dbl> <chr> 
  1 a             1 WEEK 1
2 a             2 WEEK 2
3 b             1 WEEK 1
4 b             2 WEEK 2
```
```r
> sessionInfo()
R version 4.0.5 (2021-03-31)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/atlas/libblas.so.3.10.3
LAPACK: /usr/lib/x86_64-linux-gnu/atlas/liblapack.so.3.10.3

locale:
[1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8        LC_COLLATE=C.UTF-8    
[5] LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8    LC_PAPER=C.UTF-8       LC_NAME=C             
[9] LC_ADDRESS=C           LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   

attached base packages:
  [1] stats     graphics  grDevices datasets  utils     methods   base     

other attached packages:
  [1] dplyr_1.0.5

loaded via a namespace (and not attached):
[1] tidyr_1.1.3      fansi_0.4.2      crayon_1.4.1     utf8_1.2.1       R6_2.5.0         lifecycle_1.0.3 
[7] magrittr_2.0.3   pillar_1.5.1     rlang_1.0.6      cli_3.4.1        renv_0.16.0      rstudioapi_0.13 
[13] vctrs_0.5.0      generics_0.1.0   ellipsis_0.3.2   tools_4.0.5      glue_1.6.2       purrr_0.3.4     
[19] compiler_4.0.5   pkgconfig_2.0.3  tidyselect_1.1.0 tibble_3.1.0    
> 
```

Comment (bundfussr) [reply tests/testthat/test-derive_expected_records.R]: I would check the number of columns and call `crossing()` only if there is at least one column. Once we move to tidyr 1.2.0 we can remove the work-around.

**Review by bundfussr — APPROVED** [review]

**Pull request merged**